### PR TITLE
Add file globbing search semantics to list_probes()

### PR DESCRIPTION
**Pull request by cneira**
@ajor , @brendangregg 

Checking https://github.com/iovisor/bpftrace/issues/20

I tried to improve list_probes using file globbing semantics, here are some search results.

```bash
cneira@Trixie:~/bpftrace/build$ sudo bpftrace -l '*sleep*'
tracepoint:block:block_sleeprq
tracepoint:compaction:mm_compaction_kcompactd_sleep
tracepoint:vmscan:mm_vmscan_kswapd_sleep
tracepoint:sched:sched_stat_sleep
tracepoint:syscalls:sys_enter_clock_nanosleep
tracepoint:syscalls:sys_exit_clock_nanosleep
tracepoint:syscalls:sys_enter_nanosleep
tracepoint:syscalls:sys_exit_nanosleep
kprobe:x86_acpi_enter_sleep_state
kprobe:pmc_sleep_tmr_open
kprobe:pmc_sleep_tmr_show
kprobe:wq_worker_sleeping
...
cneira@Trixie:~/bpftrace/build$ sudo bpftrace -l 'tracepoint*'
tracepoint:kvmmmu:kvm_mmu_pagetable_walk
tracepoint:kvmmmu:kvm_mmu_paging_element
tracepoint:kvmmmu:kvm_mmu_set_accessed_bit
tracepoint:kvmmmu:kvm_mmu_set_dirty_bit
tracepoint:kvmmmu:kvm_mmu_walker_error
tracepoint:kvmmmu:kvm_mmu_get_page
tracepoint:kvmmmu:kvm_mmu_sync_page
tracepoint:kvmmmu:kvm_mmu_unsync_page
tracepoint:kvmmmu:kvm_mmu_prepare_zap_page
tracepoint:kvmmmu:mark_mmio_spte
tracepoint:kvmmmu:handle_mmio_page_fault
tracepoint:kvmmmu:fast_page_fault
tracepoint:kvmmmu:kvm_mmu_invalidate_zap_all_pages
tracepoint:kvmmmu:check_mmio_spte
tracepoint:kvm:kvm_entry
tracepoint:kvm:kvm_hypercall
tracepoint:kvm:kvm_hv_hypercall
tracepoint:kvm:kvm_pio
tracepoint:kvm:kvm_fast_mmio
tracepoint:kvm:kvm_cpuid
...
cneira@Trixie:~/bpftrace/build$ sudo bpftrace -l '*:syscalls:*' 
tracepoint:syscalls:sys_enter_socket
tracepoint:syscalls:sys_exit_socket
tracepoint:syscalls:sys_enter_socketpair
tracepoint:syscalls:sys_exit_socketpair
...
cneira@Trixie:~/bpftrace/build$ sudo bpftrace -l 'tracepoint:?vm:?vm_?????call'
tracepoint:kvm:kvm_hypercall
```

### CODE REVIEW

**Comment by brendangregg** [comment]
Looks like an improvement. What does this match?: `bpftrace -l '*:sleep*'` ?

It should match probe names that start with "sleep". Based on the code, I think it'd also match "sleep" in the 2nd field for tracepoints (none exist), but I think that may be ok.

**Comment by cneira** [comment]
yes, that's correct, that will also match that string, when it's available, currently it will match

```bash
cneira@Trixie:~$ sudo bpftrace -l '*:sleep*'
kprobe:sleep_millisecs_store
kprobe:sleep_millisecs_show
```

File: src/list.cpp [hunk 1]
```diff
@@ -16,22 +16,35 @@ namespace bpftrace {
...
+}
 
-  // TODO: glob searching instead of discarding wildcards
-  for (i = 0; i < strlen(remove); ++i)
-  {
-    s.erase(std::remove(s.begin(), s.end(), remove[i]), s.end());
+bool search_probe(const std::string &probe, const std::string search)
+{
+  try {
+	std::string s = replace_all(search,"*",".*");
```
Comment (ajor) [inline src/list.cpp]: 2-space indent not tabs please!

File: src/list.cpp [hunk 2]
```diff
@@ -16,22 +16,35 @@ namespace bpftrace {
 const std::string kprobe_path = "/sys/kernel/debug/tracing/available_filter_functions";
 const std::string tp_path = "/sys/kernel/debug/tracing/events";
 
-bool search_probe(const std::string &probe, const std::string search)
+std::string replace_all(const std::string &str, const char *from, const char *to)
```
[the remark this answers is not in the document]
Comment (cneira) [reply src/list.cpp]: no problem

File: src/list.cpp [hunk 3]
```diff
@@ -16,22 +16,35 @@ namespace bpftrace {
...
+        index += to_len;
+    }
+    return result;
+}
 
-  // TODO: glob searching instead of discarding wildcards
-  for (i = 0; i < strlen(remove); ++i)
-  {
-    s.erase(std::remove(s.begin(), s.end(), remove[i]), s.end());
+bool search_probe(const std::string &probe, const std::string search)
```
[the remark this answers is not in the document]
Comment (cneira) [reply src/list.cpp]: sure

**Comment by cneira** [comment]
@ajor 
Just changed a reference to const. Thanks for reviewing!

File: src/list.cpp [hunk 4]
```diff
@@ -16,22 +16,39 @@ namespace bpftrace {
...
+std::string replace_all(const std::string &str, const std::string &from,
+			const std::string &to)
 {
-  std::string s = search;
-  char remove[] = "*.?";
-  unsigned int i;
+    std::string result(str);
+    std::string::size_type
+        index = 0,
+        from_len = strlen(from.c_str()),
```
Comment (ajor) [inline src/list.cpp]: Sorry, just spotted one more thing - this can be just `from.size()` now

Comment (cneira) [reply src/list.cpp]: you are right!.

**Comment by cneira** [comment]
@ajor 
you are right, just changed to use std::string size() method.

**Comment by brendangregg** [comment]
Thanks @cneira !

**Pull request merged**